invenio_preservation_sync/api.py

```diff
@@ -0,0 +1,163 @@
+        """Get latest Preservation Info for the given record id."""
+        preservation = (
+            PreservationInfo.query.filter_by(record_id=record_id)
+            .order_by(text(cls.query_order))
+            .first()
+        )
+        return preservation
+
+    @classmethod
+    def get_existing_preservation(
+        cls,
+        record_id=None,
+        revision_id=None,
+        archive_timestamp=None,
+    ):
+        """Return Preservation Info if it already exists."""
+        preservation = PreservationInfo.query.filter_by(
+            record_id=record_id,
+            revision_id=revision_id,
+            archive_timestamp=archive_timestamp,
+        ).first()
+        return preservation
+
+    @classmethod
+    def update_existing_preservation(
+        cls,
+        preservation,
+        status=None,
+        harvest_timestamp=None,
+        uri=None,
+        path=None,
+        event=None,
+        description=None,
+    ):
+        """Update existing preservation."""
+        status = PreservationInfoAPI.convert_status(status)
+
+        if (
+            preservation.status == status
+            and preservation.harvest_timestamp == parse(harvest_timestamp)
+            and preservation.uri == uri
+            and preservation.path == path
+            and preservation.description == description
+        ):
+            raise PreservationAlreadyReceivedError(preservation)
+
+        if status:
+            preservation.status = status
+
+        if harvest_timestamp:
+            preservation.harvest_timestamp = harvest_timestamp
+
+        if uri:
+            preservation.uri = uri
+
+        if path:
+            preservation.path = path
+
+        if description:
+            preservation.description = description
+
+        preservation.event = event
+
+        db.session.commit()
+
+    @classmethod
+    def convert_status(cls, value):
+        """Convert the status of the preservation info.
+
+        Valid examples: "P", "Preserved", "preserved"
+        """
+        if isinstance(value, str):
+            if value in [x.value for x in PreservationStatus]:
+                return PreservationStatus(value)
+            elif value.upper() in [x.name for x in PreservationStatus]:
+                return PreservationStatus[value.upper()]
+        elif isinstance(value, PreservationStatus):
+            return value
+        raise ValueError(
+            f"Invalid status value for the Preservation Status. Got: {value}"
+        )
+
+    @classmethod
+    def convert_to_dict(cls, preservation):
+        """Convert the results to dict."""
+        if isinstance(preservation, list):
+            res = []
+            for p in preservation:
+                res.append({c.name: getattr(p, c.name) for c in p.__table__.columns})
+            return res
+        return {
+            c.name: getattr(preservation, c.name)
+            for c in preservation.__table__.columns
+        }
```

Comment:
Multiple comments on this one:

1. Can you add some examples for this method? I feel this can be achieved by using `dictutils` or other utilities we use in rdm for dictionaries
2. Dynamically using all the table columns seems like a bad idea, it is safer to have a serializer and be explicit on how things are serialized.
3. If this is about serialization, perhaps it should be moved. `api.py` if data layer, it should be a wrapper around the model and not contain any sort of business or presentation layers logic.

Reply:
As discussed: Added a result class to the service instead and implemented `to_json` method which is called from the presentation layer